csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3453 @@
+
+      for (auto domain_i : c10::irange(c_tv->getRootDomain().size())) {
+        auto c_id = c_tv->getRootDomain()[domain_i];
+        auto o_id = other_tv_output->getRootDomain()[domain_i];
+        idGraph(IdMappingMode::EXACT).mapIds(o_id, c_id);
+      }
+    }
+
+    // Map producer-consumer relationships based on the root domain map
+    auto tv_inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+    for (auto p_tv : tv_inputs) {
+      // For exact mapings do not map any broadcast dimensions to
+      // non-broadcast dimensions. Prevent any broadcasted axes being mapped
+      // to non-broadcasted axes.
+      auto exact_c2p_root_map =
+          PairwiseRootDomainMap(p_tv, c_tv, true)
+              .mapConsumerToProducer(c_tv->domain(), p_tv->domain());
+
+      for (auto c_id : getSortedKeys(exact_c2p_root_map, Statement::lessThan)) {
+        auto p_id = exact_c2p_root_map.at(c_id);
+        idGraph(IdMappingMode::EXACT).mapIds(c_id, p_id);
+      }
+    }
+
+    idGraph(IdMappingMode::EXACT).mapThroughLoopSwizzles();
+  }
+}
+
+void IterDomainGraphs::buildPermissiveMap(const std::vector<Expr*>& exprs) {
+  idGraph(IdMappingMode::PERMISSIVE) = idGraph(IdMappingMode::ALMOSTEXACT);
+
+  for (auto expr : exprs) {
+    // Multiple outputs are already mapped, we can ignore all but the first
+    // consumer given they have to be replayed in the same exact way
+    // Multiple outputs are already mapped, we can ignore all but the first
+    // consumer given they have to be replayed in the same exact way
+    TensorView* c_tv = ir_utils::getTvOutput(expr);
+
+    auto tv_inputs = ir_utils::filterByType<TensorView>(expr->inputs());
+
+    for (auto p_tv : tv_inputs) {
+      auto p_ids_vec = ir_utils::allIDsOf(p_tv);
+      auto c_ids_vec = ir_utils::allIDsOf(c_tv);
+      std::unordered_set<IterDomain*> p_ids(p_ids_vec.begin(), p_ids_vec.end());
+      std::unordered_set<IterDomain*> c_ids(c_ids_vec.begin(), c_ids_vec.end());
+
+      ForwardingInfo permissive_forwarding(p_tv, c_tv);
+      for (auto entry : permissive_forwarding.producer_forwarding_map) {
+        idGraph(IdMappingMode::PERMISSIVE).mapIds(entry.first, entry.second);
+      }
+
+      // TODO: Should this just get rolled up in the forwarding map now?
+      for (auto entry : permissive_forwarding.producer_compliment_map) {
+        for (auto entry_2 : entry.second) {
+          idGraph(IdMappingMode::PERMISSIVE).mapIds(entry.first, entry_2);
+        }
+      }
+
+      for (auto entry : permissive_forwarding.consumer_forwarding_map) {
+        idGraph(IdMappingMode::PERMISSIVE).mapIds(entry.first, entry.second);
+      }
+
+      // TODO: Should this just get rolled up in the forwarding map now?
+      for (auto entry : permissive_forwarding.consumer_compliment_map) {
+        for (auto entry_2 : entry.second) {
+          idGraph(IdMappingMode::PERMISSIVE).mapIds(entry.first, entry_2);
```

Comment:
Does this mean a forwarded ID is mapped with its compliment ID? Is it the case with the current CA map as well?

Reply:
That is different and I don't remember why I did this... Maybe it's a mistake. Will mark it to look into it later.

Reply:
Just in case, it's done for the producer IDs as well.

Reply:
Thanks. I have both marked locally. I'm not really using this map at the moment, so I will check later on.